Prevents runtime issues from impacting logger creation

bkadatz opened the pull request:
Per Serilog's [Reliability](https://github.com/serilog/serilog/wiki/Reliability) page, this prevents an error in the runtime configuration from throwing an exception, which will cause the entire `new LoggerConfiguration()` fluent API calls to fail.

This change still allows invariant violations to cause a failure: when the storage connection string is invalid and cannot be parsed, it will throw a `FormatException`. However, as long as the connection string successfully parses, any issues actually connecting to the relevant Azure Storage service won't impact the ability to configure any other sinks and log to them.

bkadatz commented:
I've been thinking about this a little further and I'm thinking there should be a flag so that even a `FormatException` on the connection string doesn't impact creation of the logger. Essentially, what the file sink does with `if (propagateExceptions) throw;`. The obvious benefit is that if your connection string is in a config setting, a malformed or missing connection string won't cause all other logging to fail.

nblumhardt commented on src/Serilog.Sinks.AzureTableStorage/LoggerConfigurationAzureTableStorageExtensions.cs:
```diff
@@ -67,9 +68,20 @@ public static class LoggerConfigurationAzureTableStorageExtensions
+					(ILogEventSink)new AzureBatchingTableStorageSink(storageAccount, formatProvider, batchPostingLimit ?? DefaultBatchPostingLimit, period ?? DefaultPeriod, storageTableName, keyGenerator) :
+					new AzureTableStorageSink(storageAccount, formatProvider, storageTableName, keyGenerator);
+			}
+
```
Nit: extra newline

nblumhardt commented on src/Serilog.Sinks.AzureTableStorage/LoggerConfigurationAzureTableStorageExtensions.cs:
```diff
@@ -67,9 +68,20 @@ public static class LoggerConfigurationAzureTableStorageExtensions
+			catch (Exception ex)
+			{
+				Debugging.SelfLog.WriteLine("Error configuring AzureTableStorage: {0}", ex);
+				sink = new NullSink();
```
Not super-intuitive, but:

```csharp
sink = new LoggerConfiguration().CreateLogger();
```

avoids adding the extra `NullSink` class.

bkadatz replied:
Nice... I copied `NullSink` from the File sink project. That's a handy shortcut.

nblumhardt commented on src/Serilog.Sinks.AzureTableStorage/LoggerConfigurationAzureTableStorageWithPropertiesExtensions.cs:
```diff
@@ -70,12 +71,23 @@ public static class LoggerConfigurationAzureTableStorageWithPropertiesExtensions
+					new AzureBatchingTableStorageWithPropertiesSink(storageAccount, formatProvider, batchPostingLimit ?? DefaultBatchPostingLimit, period ?? DefaultPeriod, storageTableName, additionalRowKeyPostfix, keyGenerator)
+					: new AzureTableStorageWithPropertiesSink(storageAccount, formatProvider, storageTableName, additionalRowKeyPostfix, keyGenerator);
+			}
+
```
As above :-)

nblumhardt commented on test/Serilog.Sinks.AzureTableStorageWithProperties.Tests/AzureTableStorageWithPropertiesSinkTests.cs:
```diff
@@ -293,5 +293,54 @@ private class Struct0
+					.WriteTo.AzureTableStorage("InvalidConnectionString!!!")
+					.CreateLogger();
+
+				Log.Information("This should silently work, even when Azure Storage Emulator is not running.");
```
This line isn't needed by the test.

nblumhardt commented on test/Serilog.Sinks.AzureTableStorageWithProperties.Tests/AzureTableStorageWithPropertiesSinkTests.cs:
```diff
@@ -293,5 +293,54 @@ private class Struct0
+					.WriteTo.AzureTableStorageWithProperties("InvalidConnectionString!!!")
+					.CreateLogger();
+
+				Log.Information("This should silently work, even when Azure Storage Emulator is not running.");
```
As above.

nblumhardt commented:
Thank you for the PR - looks good! Just a few minor comments.

bkadatz commented:
I fixed these issues and went ahead and added the `propagateExceptions` parameter so that users can choose whether or not exceptions are thrown. The code defaults to not throwing exceptions and only logging the failure in Serilog's `SelfLog` output.

nblumhardt commented:
Thanks for the update! At this stage, I'd rather not add `propagateExceptions` as a parameter, since something like this should be consistent across sinks. Can we just assume `false` for runtime issues?

Otherwise 👍 :shipit:

bkadatz commented:
Done.

nblumhardt commented on src/Serilog.Sinks.AzureTableStorage/LoggerConfigurationAzureTableStorageExtensions.cs:
```diff
@@ -102,8 +114,19 @@ public static class LoggerConfigurationAzureTableStorageExtensions
+            try
+            {
+                var storageAccount = CloudStorageAccount.Parse(connectionString);
+                return AzureTableStorage(loggerConfiguration, storageAccount, restrictedToMinimumLevel, formatProvider, storageTableName, writeInBatches, period, batchPostingLimit, keyGenerator);
```
This is a call to the other method that already contains a try/catch block; can we skip the extra try/catch here?

bkadatz replied:
Both are needed to prevent runtime exceptions as we don't know which constructor will be used. The try/catch here catches any exceptions from the constructor that parses the connection string into a `CloudStorageAccount` variable. However, end users may also call the constructor with an already-created `CloudStorageAccount` instance directly, so a try/catch is needed there as well (as a communication failure will raise an exception).
